### PR TITLE
Implement a request to sign urls.

### PR DESCRIPTION
**Pull request by coryan**
Though signing a URL is a local operation, passing options to the Client
member function is convenient and consistent with the other functions.
This PR introduces a request to capture all the parameters to a signed
URL, including the variable number of extension headers.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1575)

### CODE REVIEW

File: google/cloud/storage/internal/signed_url_requests.cc [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+  os << verb() << "\n"
+     << md5_hash_value_ << "\n"
+     << content_type_ << "\n"
+     << seconds.count() << "\n";
+
+  for (auto const& kv : extension_headers_) {
+    os << kv.first << ":" << kv.second << "\n";
+  }
+  os << "/" << bucket_name() << "/" << object_name();
```
Comment (houglum) [inline google/cloud/storage/internal/signed_url_requests.cc]: We should be URL-encoding the object name here, as done in the Python client lib:
https://github.com/googleapis/google-cloud-python/blob/6b327e9eedae1430c4abe2a628faeb1f96bc590a/storage/google/cloud/storage/blob.py#L352

The Python library uses the `quote` method from `urllib`, which is almost the same as `curl_easy_escape`... except that `curl_easy_escape` also URL-encodes forward slashes.  I think the correct behavior is actually to escape the forward slashes anyway, so curl's version should suffice.

File: google/cloud/storage/internal/signed_url_requests.h [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      : verb_(std::move(verb)),
+        bucket_name_(std::move(bucket_name)),
+        object_name_(std::move(object_name)) {}
+
+  std::string const& verb() const { return verb_; }
+  std::string const& bucket_name() const { return bucket_name_; }
+  std::string const& object_name() const { return object_name_; }
+
+  /// Creates the blob to be signed.
+  std::string CreateBlob() const;
```
Comment (houglum) [inline google/cloud/storage/internal/signed_url_requests.h]: `CreateBlob()` sounds like something that would upload content or create a file.  Can we name this something like `StringToSign()`?

File: google/cloud/storage/internal/signed_url_requests_test.cc [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  // Used this command to get the date in seconds:
+  //   date +%s -u --date=2018-12-03T12:00:00Z
+  std::string expected_blob = R"""(GET
+rmYdCNHKFXam78uCt7xQLw==
+text/plain
+1543838400
+x-goog-acl:project-private
+x-goog-encryption-algorithm:AES256
+x-goog-meta-foo:bar,baz
+/test-bucket/test-object)""";
```
Comment (houglum) [inline google/cloud/storage/internal/signed_url_requests_test.cc]: If we decide to test with an object that has a forward slash in the name, can we make sure here that `my/object:name` shows up as `/test-bucket/my%2Fobject%3Aname`?

File: google/cloud/storage/internal/signed_url_requests.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/signed_url_requests.cc]: The documentation:

https://cloud.google.com/storage/docs/access-control/signed-urls#about-canonical-resources

suggests that forward slashes should be preserved.  Let me test this and get back to you.

File: google/cloud/storage/internal/signed_url_requests.h (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/signed_url_requests.h]: Done.

File: google/cloud/storage/internal/signed_url_requests_test.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/internal/signed_url_requests_test.cc]: N/A ? Since there is not need to URL-escape the resource in the signature block.

**Comment by coryan** [comment]
PTAL.

File: google/cloud/storage/internal/signed_url_requests.cc (hunk 1 above)
Comment (houglum) [reply google/cloud/storage/internal/signed_url_requests.cc]: There are a number of other characters that are valid to use in object names, but could potentially cause problems in URLs (e.g. the space character, question mark, and ampersand come to mind).

Comment (coryan) [reply google/cloud/storage/internal/signed_url_requests.cc]: Doh!  Indeed.  Thanks for your patience, fixed.

File: google/cloud/storage/internal/signed_url_requests_test.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/internal/signed_url_requests_test.cc]: Done.

**Comment by coryan** [comment]
PTAL.

**Pull request merged**